File: problems/array/0560-subarray-sum-equals-k.js
/**
 * #560 Subarray Sum Equals K
 *
 * Link: https://leetcode.com/problems/subarray-sum-equals-k/
 * Tags: array, prefix sum, hash map
 * Difficulty: Medium
 *
 * If a current prefix sum is `sum`, every earlier prefix sum of `sum - k`
 * leaves a subarray with sum `k`. The map stores the count because identical
 * earlier prefix sums represent distinct starting positions.
 *
 * Time complexity: O(n)
 * Space complexity: O(n)
 *
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var subarraySum = function (nums, k) {
  const prefixFrequencies = new Map([[0, 1]])
  let prefixSum = 0
  let count = 0

  for (const number of nums) {
    prefixSum += number
    count += prefixFrequencies.get(prefixSum - k) ?? 0
    prefixFrequencies.set(
      prefixSum,
      (prefixFrequencies.get(prefixSum) ?? 0) + 1,
    )
  }

  return count
}

File: problems/array/0560-subarray-sum-equals-k.py
"""#560 Subarray Sum Equals K
Count earlier prefix sums equal to current_prefix - k.
Time: O(n) | Space: O(n)
"""


class Solution:
    def subarraySum(self, nums, k):
        frequencies = {0: 1}
        prefix_sum = 0
        total = 0

        for number in nums:
            prefix_sum += number
            total += frequencies.get(prefix_sum - k, 0)
            frequencies[prefix_sum] = frequencies.get(prefix_sum, 0) + 1

        return total
